pkg/applier/sharded.go

```diff
@@ -0,0 +1,808 @@
+
+			// Check if all chunklets for this work are complete
+			if pending.completedChunklets == pending.totalChunklets {
+				a.logger.Debug("feedbackCoordinator all chunklets complete, invoking callback", "workID", completion.workID)
+
+				// Invoke the callback
+				callback := pending.callback
+				affectedRows := pending.totalAffectedRows
+
+				// Remove completed work from pending map
+				delete(a.pendingWork, completion.workID)
+				a.pendingMutex.Unlock()
+
+				// Invoke callback outside the lock
+				callback(affectedRows, nil)
+			} else {
+				a.pendingMutex.Unlock()
+			}
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// DeleteKeys deletes rows by their key values synchronously, broadcasting to all shards.
+// The keys are hashed key strings (from utils.HashKey).
+// If lock is non-nil, operations are executed under table locks (one per shard).
+//
+// Note: we only track modifications by PRIMARY KEY, not by shard key (aka primary vindex).
+// For this reason we can't extract the vindex value, and must instead broadcast
+// the deletes to all shards. The vindex value is considered immutable, and we will
+// error if it changes on an update.
+func (a *ShardedApplier) DeleteKeys(ctx context.Context, sourceTable, targetTable *table.TableInfo, keys []string, lock *dbconn.TableLock) (int64, error) {
+	if len(keys) == 0 {
+		return 0, nil
+	}
+
+	// In move operations, targetTable may be nil because there isn't a single target table.
+	// In this case, use sourceTable since the schema is the same.
+	if targetTable == nil {
+		targetTable = sourceTable
+	}
+
+	// Convert hashed keys to row value constructor format
+	var pkValues []string
+	for _, key := range keys {
+		pkValues = append(pkValues, utils.UnhashKeyToString(key))
+	}
+
+	// Build DELETE statement
+	// Use just the table name, not the fully qualified name, because
+	// the database connection (shard.writeDB) already determines which database to write to
+	tableName := fmt.Sprintf("`%s`", targetTable.TableName)
+	deleteStmt := fmt.Sprintf("DELETE FROM %s WHERE (%s) IN (%s)",
+		tableName,
+		table.QuoteColumns(sourceTable.KeyColumns),
+		strings.Join(pkValues, ","),
+	)
+
+	// Execute deletes on all shards in parallel (broadcast)
+	var totalAffected int64
+	var mu sync.Mutex
+	var errGroup error
+
+	var wg sync.WaitGroup
+	for _, shard := range a.shards {
+		wg.Add(1)
+		go func(shard *shardTarget) {
+			defer wg.Done()
+
+			var affected int64
+			var err error
+			// Execute under lock if provided
+			if lock != nil {
+				if err = lock.ExecUnderLock(ctx, deleteStmt); err != nil {
+					err = fmt.Errorf("failed to execute delete under lock on shard %d: %w", shard.shardID, err)
+				} else {
+					// We can't know the actual affected rows when using lock, so estimate
+					affected = 0
+				}
+			} else {
+				// Execute as a retryable transaction
+				affected, err = dbconn.RetryableTransaction(ctx, shard.writeDB, false, shard.dbConfig, deleteStmt)
```

Comment:
Is this using the same delete statement on all shards for expediency, or would we actually expect the same keys to be on multiple shards somehow?